pkg/querysharding/analyzer.go

```diff
@@ -71,14 +122,34 @@ func (a *QueryAnalyzer) Analyze(query string) (QueryAnalysis, error) {
 	})
 
 	if !isShardable {
-		return nonShardableQuery(), nil
+
+		if ok := LRU.Add(query, cachedValue{
+			QueryAnalysis: nonShardableQuery,
+			err:           nil,
+		}); ok {
+			logger.Log("cachingError=", query)
+		}
+		return nonShardableQuery, nil
 	}
 
 	rootAnalysis := analyzeRootExpression(expr)
 	if rootAnalysis.IsShardable() && rootAnalysis.shardBy {
+		if ok := LRU.Add(query, cachedValue{
+			QueryAnalysis: rootAnalysis,
+			err:           nil,
+		}); ok {
+			logger.Log("cachingError=", query)
+		}
 		return rootAnalysis, nil
 	}
 
+	if ok := LRU.Add(query, cachedValue{
+		QueryAnalysis: analysis,
+		err:           nil,
+	}); ok {
+		logger.Log("cachingError=", query)
+	}
+
 	return analysis, nil
```

Comment:
Since we need to cache results and error so the code overall is kind of messy now. What about making `Analyze` an interface and another struct like `CachedQueryAnalyzer`? Basically the flow would be:

```go
func (a *CachedQueryAnalyzer) Analyze(query string) (QueryAnalysis, error) {
    if a.cache.Exists(query) { return a.cache.Get(query) }

    // Analyze if needed
    analysis, err := a.Analyze(query)
    
    // Add to cache
    a.cache.Add(query, xxx)
    return ...
}
```